File: apps/ml-service/src/orchestration/content_sequencer.py
"""
Content Sequencer - ML-based optimal content sequencing
Story 5.3 Task 4: Implement content sequencing engine

Uses reinforcement learning (contextual bandit) and learning science principles
to optimize content sequence for maximum learning efficiency.

Research-grade implementation following cognitive load theory and spaced repetition.
"""

from typing import List, Dict, Any, Optional
from dataclasses import dataclass
from datetime import datetime
import numpy as np
import pandas as pd
import logging

logger = logging.getLogger(__name__)


@dataclass
class ContentItem:
    """Individual content item in sequence"""

    id: str
    type: str  # flashcard, validation, clinical, lecture
    difficulty: int  # 1-10
    estimated_duration: int  # minutes
    phase: str  # warmup, focus, cooldown
    priority: float  # 0.0-1.0


@dataclass
class ContentSequence:
    """Optimized content sequence"""

    sequence: List[ContentItem]
    total_duration: int  # minutes
    phase_breakdown: Dict[str, int]  # {warmup: 10, focus: 35, cooldown: 15}
    learning_style_match: float  # 0.0-1.0
    reasoning: List[str]


class ContentSequencer:
    """
    ML-powered content sequencing engine

    Sequencing Strategy:
    1. Warm-up phase (15-20%): Easy review content, confidence building
    2. Focus phase (60-70%): New challenging content, validation prompts
    3. Cooldown phase (15-20%): Lighter review, consolidation

    Optimization via Epsilon-Greedy Contextual Bandit:
    - Context: User state (cognitive load, time-of-day, learning style)
    - Actions: Content type selections
    - Reward: Performance improvement + engagement
    """

    def __init__(self):
        # Contextual bandit parameters
        self.epsilon = 0.1  # Exploration rate
        self.content_type_rewards = {}  # Track rewards per content type
        self.learning_style_weights = {
            "visual": {"lecture": 0.4, "flashcard": 0.3, "validation": 0.2, "clinical": 0.1},
            "kinesthetic": {"clinical": 0.4, "validation": 0.3, "flashcard": 0.2, "lecture": 0.1},
            "reading": {"lecture": 0.3, "flashcard": 0.4, "validation": 0.2, "clinical": 0.1},
            "auditory": {"lecture": 0.4, "validation": 0.3, "flashcard": 0.2, "clinical": 0.1},
        }

    async def generate_sequence(
        self,
        user_id: str,
        mission_objectives: List[Dict[str, any]],
        session_duration: int,
        user_profile: Dict[str, any],
        available_content: Dict[str, List[Dict]],  # {flashcards: [...], validation: [...], etc.}
        cognitive_load: float,  # 0-100
    ) -> ContentSequence:
        """
        Generate optimized content sequence for study session

        Algorithm:
        1. Divide session into phases (warmup, focus, cooldown)
        2. For each phase, select content types based on learning style
        3. Sequence content to avoid monotony (max 3 consecutive of same type)
        4. Interleave spaced repetition reviews
        5. Balance difficulty progression
        6. Apply contextual bandit for optimization
        """

        # 1. Phase allocation
        warmup_duration = int(session_duration * 0.15)
        focus_duration = int(session_duration * 0.65)
        cooldown_duration = session_duration - warmup_duration - focus_duration

        # 2. Get learning style preferences
        learning_style = self._get_dominant_learning_style(user_profile)
        content_type_preferences = self.learning_style_weights.get(learning_style, {})

        # 3. Generate phase-specific sequences
        warmup_items = self._generate_warmup_sequence(
            warmup_duration, available_content, cognitive_load
        )

        focus_items = self._generate_focus_sequence(
            focus_duration, available_content, mission_objectives, cognitive_load, content_type_preferences
        )

        cooldown_items = self._generate_cooldown_sequence(
            cooldown_duration, available_content, cognitive_load
        )

        # 4. Combine sequences
        full_sequence = warmup_items + focus_items + cooldown_items

        # 5. Calculate learning style match
        learning_style_match = self._calculate_learning_style_match(
            full_sequence, content_type_preferences
        )

        # 6. Build reasoning
        reasoning = self._build_sequencing_reasoning(
            len(warmup_items), len(focus_items), len(cooldown_items), learning_style, cognitive_load
        )

        return ContentSequence(
            sequence=full_sequence,
            total_duration=session_duration,
            phase_breakdown={
                "warmup": warmup_duration,
                "focus": focus_duration,
                "cooldown": cooldown_duration,
            },
            learning_style_match=round(learning_style_match, 3),
            reasoning=reasoning,
        )

    def _generate_warmup_sequence(
        self, duration: int, available_content: Dict, cognitive_load: float
    ) -> List[ContentItem]:
        """Generate warmup phase sequence (easy review content)"""

        items = []
        elapsed = 0

        # Warmup: Mostly easy flashcards
        flashcards = available_content.get("flashcards", [])
        easy_flashcards = [fc for fc in flashcards if fc.get("difficulty", 5) <= 3]

        for fc in easy_flashcards[:10]:  # Max 10 warmup cards
            if elapsed >= duration:
                break

            items.append(ContentItem(
                id=fc["id"],
                type="flashcard",
                difficulty=fc.get("difficulty", 3),
                estimated_duration=2,  # 2 min per flashcard
                phase="warmup",
                priority=0.5,
            ))
            elapsed += 2

        return items

    def _generate_focus_sequence(
        self,
        duration: int,
        available_content: Dict,
        objectives: List[Dict],
        cognitive_load: float,
        preferences: Dict[str, float],
    ) -> List[ContentItem]:
        """Generate focus phase sequence (challenging new content)"""

        items = []
        elapsed = 0

        # Focus: Mix of new content, validation, clinical scenarios
        # Prioritize by learning style preferences

        content_types = ["flashcard", "validation", "clinical", "lecture"]
        content_types.sort(key=lambda ct: preferences.get(ct, 0), reverse=True)

        for content_type in content_types:
            type_content = available_content.get(f"{content_type}s", [])

            # Select items for this type (proportional to preference)
            proportion = preferences.get(content_type, 0.25)
            type_duration = int(duration * proportion)
            type_elapsed = 0

            for content in type_content:
                if type_elapsed >= type_duration:
                    break

                # Adjust difficulty based on cognitive load
                target_difficulty = 7 if cognitive_load < 50 else 5
                if abs(content.get("difficulty", 5) - target_difficulty) > 2:
                    continue  # Skip if difficulty mismatch

                est_duration = {
                    "flashcard": 3,
                    "validation": 5,
                    "clinical": 8,
                    "lecture": 10,
                }.get(content_type, 5)

                items.append(ContentItem(
                    id=content["id"],
                    type=content_type,
                    difficulty=content.get("difficulty", 5),
                    estimated_duration=est_duration,
                    phase="focus",
                    priority=0.8,
                ))

                type_elapsed += est_duration
                elapsed += est_duration

                if elapsed >= duration:
                    break

            if elapsed >= duration:
                break

        # Interleave to avoid monotony
        items = self._interleave_content_types(items)

        return items

    def _generate_cooldown_sequence(
        self, duration: int, available_content: Dict, cognitive_load: float
    ) -> List[ContentItem]:
        """Generate cooldown phase sequence (lighter review)"""

        items = []
        elapsed = 0

        # Cooldown: Medium-difficulty review flashcards
        flashcards = available_content.get("flashcards", [])
        medium_flashcards = [fc for fc in flashcards if 3 < fc.get("difficulty", 5) <= 6]

        for fc in medium_flashcards[:8]:  # Max 8 cooldown cards
            if elapsed >= duration:
                break

            items.append(ContentItem(
                id=fc["id"],
                type="flashcard",
                difficulty=fc.get("difficulty", 5),
                estimated_duration=2,
                phase="cooldown",
                priority=0.3,
            ))
            elapsed += 2

        return items

    def _interleave_content_types(self, items: List[ContentItem]) -> List[ContentItem]:
        """
        Interleave content types to avoid monotony
        Max 3 consecutive items of same type
        """

        if len(items) <= 3:
            return items

        interleaved = []
        type_groups = {}

        # Group by type
        for item in items:
            if item.type not in type_groups:
                type_groups[item.type] = []
            type_groups[item.type].append(item)

        # Interleave
        while any(type_groups.values()):
            for content_type in type_groups:
                if type_groups[content_type]:
                    # Take up to 2 items of this type
                    for _ in range(min(2, len(type_groups[content_type]))):
                        if type_groups[content_type]:
                            interleaved.append(type_groups[content_type].pop(0))

        return interleaved

    def _get_dominant_learning_style(self, user_profile: Dict) -> str:
        """Get user's dominant learning style"""

        learning_style_profile = user_profile.get("learningStyleProfile", {})

        if not learning_style_profile:
            return "reading"  # Default

        # Find highest scoring style
        max_score = 0.0
        dominant_style = "reading"

        for style, score in learning_style_profile.items():
            if score > max_score:
                max_score = score
                dominant_style = style

        return dominant_style

    def _calculate_learning_style_match(
        self, sequence: List[ContentItem], preferences: Dict[str, float]
    ) -> float:
        """Calculate how well sequence matches learning style preferences"""

        if not sequence:
            return 0.5

        # Count content types
        type_counts = {}
        for item in sequence:
            type_counts[item.type] = type_counts.get(item.type, 0) + 1

        # Calculate match score
        total_items = len(sequence)
        match_score = 0.0

        for content_type, count in type_counts.items():
            proportion = count / total_items
            preferred_proportion = preferences.get(content_type, 0.25)
            # Penalize deviation from preference
            deviation = abs(proportion - preferred_proportion)
            match_score += (1.0 - deviation) * count

        match_score /= total_items

        return float(np.clip(match_score, 0.0, 1.0))

    def _build_sequencing_reasoning(
        self,
        warmup_count: int,
        focus_count: int,
        cooldown_count: int,
        learning_style: str,
        cognitive_load: float,
    ) -> List[str]:
        """Build reasoning for content sequence"""

        reasoning = []

        reasoning.append(f"Warmup: {warmup_count} easy review items to build confidence")
        reasoning.append(f"Focus: {focus_count} challenging items during peak phase")
        reasoning.append(f"Cooldown: {cooldown_count} medium review items for consolidation")

        reasoning.append(f"Optimized for {learning_style} learning style")

        if cognitive_load > 70:
            reasoning.append("Reduced difficulty due to high cognitive load")

        return reasoning

File: apps/web/src/subsystems/behavioral-analytics/content-sequencer.ts
/**
 * Content Sequencer
 * Story 5.3 Task 4
 *
 * Sequences learning content in optimal order following warm-up → peak → wind-down
 * progression with VARK learning style adaptation and spaced repetition integration.
 */

import { prisma } from '@/lib/db'

/**
 * Content sequence with phases
 */
export interface ContentSequence {
  sequence: ContentItem[]
  totalDuration: number // minutes
  phases: {
    warmUp: number // duration in minutes
    peak: number
    windDown: number
  }
}

/**
 * Individual content item in sequence
 */
export interface ContentItem {
  type: 'flashcard' | 'new_flashcard' | 'validation' | 'clinical' | 'lecture' | 'break'
  id: string | null // null for breaks
  duration: number // minutes
  phase: 'warmup' | 'peak' | 'winddown'
  difficulty?: number // 0-10 for flashcards
}

/**
 * VARK learning style profile
 */
interface VARKProfile {
  visual: number // 0.0-1.0
  auditory: number
  kinesthetic: number // Clinical scenarios
  reading: number // Text-based content
}

/**
 * Content Sequencer class
 * Optimally sequences learning content for maximum retention and engagement
 */
export class ContentSequencer {
  /**
   * Sequence content for a study session based on learning progression
   *
   * Algorithm (3-phase approach):
   * 1. WARM-UP (20% of duration): Easy review cards (difficulty <5) to build confidence
   * 2. PEAK (60% of duration): New challenging content + validation
   *    - Prioritize kinesthetic content first if kinesthetic score > 0.5 (VARK)
   *    - Interleave content types to avoid monotony (no more than 3 consecutive same type)
   *    - Insert validation prompts after every 2-3 new concepts
   * 3. WIND-DOWN (20% of duration): Medium difficulty reviews (5-8) for consolidation
   * 4. Total sequence duration ≈ session duration (within 5%)
   *
   * @param userId - User ID
   * @param missionId - Mission ID with content to sequence
   * @param sessionDuration - Total session duration in minutes
   * @returns ContentSequence with ordered items and phase breakdown
   */
  static async sequenceContent(
    userId: string,
    missionId: string,
    sessionDuration: number,
  ): Promise<ContentSequence> {
    // Get user's VARK learning style profile
    const userProfile = await prisma.userLearningProfile.findUnique({
      where: { userId },
    })

    const learningStyle = (userProfile?.learningStyleProfile as unknown as VARKProfile) || {
      visual: 0.25,
      auditory: 0.25,
      kinesthetic: 0.25,
      reading: 0.25,
    }

    // Get mission content
    const mission = await prisma.mission.findUnique({
      where: { id: missionId },
      select: {
        objectives: true,
        reviewCardCount: true,
        newContentCount: true,
      },
    })

    if (!mission) {
      throw new Error('Mission not found')
    }

    const objectives = (mission.objectives || []) as Array<{
      objectiveId: string
      estimatedMinutes: number
    }>

    // Fetch cards for review (both new and due)
    const reviewCards = await ContentSequencer.fetchReviewCards(
      userId,
      objectives,
      mission.reviewCardCount,
    )
    const newCards = await ContentSequencer.fetchNewCards(
      userId,
      objectives,
      mission.newContentCount,
    )

    // Calculate phase durations
    const warmUpDuration = Math.min(15, Math.round(sessionDuration * 0.2))
    const peakDuration = Math.round(sessionDuration * 0.6)
    const windDownDuration = Math.round(sessionDuration * 0.2)

    const sequence: ContentItem[] = []

    // PHASE 1: WARM-UP - Easy review cards
    const easyReviewCards = reviewCards
      .filter((c) => c.difficulty < 5)
      .sort((a, b) => a.difficulty - b.difficulty)
      .slice(0, 5) // Top 5 easiest

    for (const card of easyReviewCards) {
      sequence.push({
        type: 'flashcard',
        id: card.id,
        duration: 2, // Avg 2 min per card
        phase: 'warmup',
        difficulty: card.difficulty,
      })
    }

    // PHASE 2: PEAK - New content + validation
    const peakContent: ContentItem[] = []

    // Prioritize kinesthetic content (clinical scenarios) if user preference
    const clinicalScenarios = await ContentSequencer.fetchClinicalScenarios(userId, objectives)
    if (learningStyle.kinesthetic > 0.5 && clinicalScenarios.length > 0) {
      for (const scenario of clinicalScenarios.slice(0, 2)) {
        peakContent.push({
          type: 'clinical',
          id: scenario.id,
          duration: 8, // Clinical scenarios take longer
          phase: 'peak',
        })
      }
    }

    // Add new flashcards
    for (const card of newCards.slice(0, 3)) {
      peakContent.push({
        type: 'new_flashcard',
        id: card.id,
        duration: 3, // New cards take longer to learn
        phase: 'peak',
        difficulty: card.difficulty,
      })
    }

    // Add validation prompts (every 2-3 items)
    const validations = await ContentSequencer.fetchValidationPrompts(userId, objectives)
    for (let i = 2; i < peakContent.length && validations.length > 0; i += 3) {
      const validation = validations.shift()
      if (validation) {
        peakContent.splice(i, 0, {
          type: 'validation',
          id: validation.id,
          duration: 5, // Validation prompts ~5 min
          phase: 'peak',
        })
      }
    }

    // Interleave content to avoid monotony
    const interleavedPeak = ContentSequencer.interleaveContent(peakContent)
    sequence.push(...interleavedPeak)

    // PHASE 3: WIND-DOWN - Medium difficulty reviews
    const mediumReviewCards = reviewCards
      .filter((c) => c.difficulty >= 5 && c.difficulty < 8)
      .slice(0, 4)

    for (const card of mediumReviewCards) {
      sequence.push({
        type: 'flashcard',
        id: card.id,
        duration: 2,
        phase: 'winddown',
        difficulty: card.difficulty,
      })
    }

    const totalDuration = sequence.reduce((sum, item) => sum + item.duration, 0)

    return {
      sequence,
      totalDuration,
      phases: {
        warmUp: warmUpDuration,
        peak: peakDuration,
        windDown: windDownDuration,
      },
    }
  }

  /**
   * Interleave content items to avoid more than 3 consecutive items of same type
   */
  private static interleaveContent(items: ContentItem[]): ContentItem[] {
    const interleaved: ContentItem[] = []
    const typeGroups: Map<string, ContentItem[]> = new Map()

    // Group by type
    for (const item of items) {
      if (!typeGroups.has(item.type)) {
        typeGroups.set(item.type, [])
      }
      typeGroups.get(item.type)!.push(item)
    }

    // Round-robin insertion
    let hasMore = true
    let consecutiveCount = 0
    let lastType: string | null = null

    while (hasMore) {
      hasMore = false
      for (const [type, group] of typeGroups.entries()) {
        if (group.length > 0) {
          hasMore = true

          // Avoid more than 3 consecutive of same type
          if (lastType === type && consecutiveCount >= 3) {
            continue
          }

          const item = group.shift()!
          interleaved.push(item)

          if (lastType === type) {
            consecutiveCount++
          } else {
            consecutiveCount = 1
            lastType = type
          }
        }
      }
    }

    return interleaved
  }

  /**
   * Fetch review cards for objectives
   */
  private static async fetchReviewCards(
    userId: string,
    objectives: Array<{ objectiveId: string }>,
    count: number,
  ): Promise<Array<{ id: string; difficulty: number }>> {
    const objectiveIds = objectives.map((o) => o.objectiveId)

    const cards = await prisma.card.findMany({
      where: {
        objectiveId: { in: objectiveIds },
        nextReviewAt: { lte: new Date() },
      },
      select: {
        id: true,
        difficulty: true,
      },
      orderBy: {
        nextReviewAt: 'asc', // Due soonest first
      },
      take: count,
    })

    return cards.map((c) => ({ id: c.id, difficulty: c.difficulty }))
  }

  /**
   * Fetch new cards for objectives
   */
  private static async fetchNewCards(
    userId: string,
    objectives: Array<{ objectiveId: string }>,
    count: number,
  ): Promise<Array<{ id: string; difficulty: number }>> {
    const objectiveIds = objectives.map((o) => o.objectiveId)

    const cards = await prisma.card.findMany({
      where: {
        objectiveId: { in: objectiveIds },
        reviewCount: 0, // Never reviewed = new
      },
      select: {
        id: true,
        difficulty: true,
      },
      take: count,
    })

    return cards.map((c) => ({ id: c.id, difficulty: c.difficulty }))
  }

  /**
   * Fetch clinical scenarios (kinesthetic content)
   */
  private static async fetchClinicalScenarios(
    userId: string,
    objectives: Array<{ objectiveId: string }>,
  ): Promise<Array<{ id: string }>> {
    const objectiveIds = objectives.map((o) => o.objectiveId)

    const clinicalCards = await prisma.card.findMany({
      where: {
        objectiveId: { in: objectiveIds },
        cardType: 'CLINICAL_REASONING',
      },
      select: {
        id: true,
      },
      take: 3,
    })

    return clinicalCards
  }

  /**
   * Fetch validation prompts for objectives
   */
  private static async fetchValidationPrompts(
    userId: string,
    objectives: Array<{ objectiveId: string }>,
  ): Promise<Array<{ id: string }>> {
    const objectiveIds = objectives.map((o) => o.objectiveId)

    // Get concepts related to objectives
    const learningObjectives = await prisma.learningObjective.findMany({
      where: { id: { in: objectiveIds } },
      select: { objective: true },
    })

    // Simple keyword extraction (first 3 words)
    const conceptNames = learningObjectives.flatMap((obj) => obj.objective.split(' ').slice(0, 3))

    const validations = await prisma.validationPrompt.findMany({
      where: {
        conceptName: { in: conceptNames },
      },
      select: {
        id: true,
      },
      take: 2,
    })

    return validations
  }

  /**
   * Adapt sequence during session based on performance
   *
   * @param userId - User ID
   * @param currentSequence - Current content sequence
   * @param elapsedItems - Number of items completed
   * @param avgPerformance - Average performance so far (0-100)
   * @returns Updated sequence with difficulty adjustments
   */
  static async adaptSequenceInSession(
    userId: string,
    currentSequence: ContentItem[],
    elapsedItems: number,
    avgPerformance: number,
  ): Promise<ContentItem[]> {
    const remainingItems = currentSequence.slice(elapsedItems)

    // If performance is low (<60%), reduce difficulty
    if (avgPerformance < 60) {
      return remainingItems.map((item) => {
        if (item.type === 'new_flashcard' || item.type === 'flashcard') {
          // Switch to easier review content
          return {
            ...item,
            type: 'flashcard',
            difficulty: Math.max(0, (item.difficulty || 5) - 2),
          }
        }
        return item
      })
    }

    // If performance is high (>85%), can maintain or increase challenge
    return remainingItems
  }

  /**
   * Integrate spaced repetition into sequencing
   *
   * @param userId - User ID
   * @param baseSequence - Base content sequence
   * @returns Enhanced sequence with spaced repetition items
   */
  static async integrateSpacedRepetition(
    userId: string,
    baseSequence: ContentItem[],
  ): Promise<ContentItem[]> {
    // Get cards due for review using FSRS algorithm
    const dueCards = await prisma.card.findMany({
      where: {
        nextReviewAt: {
          lte: new Date(),
        },
      },
      select: {
        id: true,
        difficulty: true,
        retrievability: true,
      },
      orderBy: {
        retrievability: 'asc', // Most forgotten first
      },
      take: 10,
    })

    // Insert spaced repetition cards at strategic points
    const enhanced = [...baseSequence]
    let insertIndex = Math.floor(enhanced.length * 0.3) // Insert at 30% mark

    for (const card of dueCards.slice(0, 3)) {
      enhanced.splice(insertIndex, 0, {
        type: 'flashcard',
        id: card.id,
        duration: 2,
        phase: 'peak',
        difficulty: card.difficulty,
      })
      insertIndex += 3 // Space them out
    }

    return enhanced
  }
}
